File: src/main/python/easy/Q0145.py
"""
145. Binary Tree Postorder Traversal
https://leetcode.com/problems/binary-tree-postorder-traversal/
"""
from __future__ import annotations
import dataclasses


@dataclasses.dataclass
class TreeNode:
    val: int
    left: TreeNode
    right: TreeNode


# Time: O(n), Space: O(nlog n), where n is the number of nodes in the tree.
def postorder_traversal(root: TreeNode) -> list[int]:
    if not root:
        return []
    l: list[int] = []

    def postorder(node: TreeNode) -> None:
        if not node.left and not node.right:
            l.append(node.val)
            return
        if node.left:
            postorder(node.left)
        if node.right:
            postorder(node.right)
        l.append(node.val)

    postorder(root)
    return l

File: src/main/cpp/easy/Q0145.cpp
// 145. Binary Tree Postorder Traversal
// https://leetcode.com/problems/binary-tree-postorder-traversal/
#include <iostream>
#include <vector>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int i) : val(i), left(nullptr), right(nullptr) {}
};

// Time: O(n), Space: O(nlog n), where n is the number of nodes in the tree.
auto postorder_traversal(TreeNode *root) -> std::vector<int> {
    if (!root) return {};

    std::vector<int> v;

    std::function<void(TreeNode *)> traverse = [&traverse, &v](const TreeNode *node) -> void {
        if (!node) throw std::invalid_argument("node cannot be null");

        if (!node->left and !node->right) {
            v.push_back(node->val);
            return;
        }
        if (node->left) {
            traverse(node->left);
        }
        if (node->right) {
            traverse(node->right);
        }
        v.push_back(node->val);
    };

    traverse(root);
    return v;
}
